integration-tests/tests/utils/port_utils.py

```diff
@@ -0,0 +1,121 @@
+"""Functions for facilitating the port connections for the CLP package."""
+
+import socket
+from typing import Any
+
+from clp_py_utils.clp_config import ClpConfig
+
+# Port constants.
+MIN_NON_PRIVILEGED_PORT = 1024
+MAX_PORT = 65535
+VALID_PORT_RANGE = range(MIN_NON_PRIVILEGED_PORT, MAX_PORT + 1)
+
+# CLP constants.
+REDUCER_MAX_PORTS = 128
+PORT_LIKE_ATTR_NAMES = [
+    "port",
+    "base_port",
+]
+
+
+def assign_ports_from_base(base_port: int, clp_config: ClpConfig) -> None:
+    """
+    Assign ports for all components described in `clp_config` that require a port. Ports are
+    assigned relative to `base_port`.
+
+    :param base_port:
+    :param clp_config:
+    :raise ValueError: if the base port is out of range or if any required port is in use.
+    """
+    # Discover which components in ClpConfig have a port-like data member.
+    component_port_targets: list[tuple[Any, str, int]] = []
+    for attr_name, attr_value in vars(clp_config).items():
+        if attr_name.startswith("_") or attr_value is None:
+            continue
```

Comment:
I think we can leverage [get_required_component_list](https://github.com/y-scope/clp/blob/fa7f5500998929030dfaacfdd43d88e676b0992b/integration-tests/tests/utils/clp_mode_utils.py#L91) to save repeated work of iterating through all components again.